src/data_chain/mod.rs

```diff
@@ -35,25 +35,18 @@ use std::collections::BTreeSet;
 /// Calculates whether a quorum of nodes have voted.  In this case, "quorum" means >50% of the
 /// members in `valid_nodes_itr` are included in `voters_itr` and that their cumulative age is >50%
 /// of the cumulative age of all members of `valid_nodes_itr`.
-pub fn quorum<'a, 'b, I, J>(voters_itr: I, valid_nodes_itr: J) -> bool
-where
-    I: IntoIterator<Item = &'a PublicInfo>,
-    J: IntoIterator<Item = &'b PublicInfo>,
-{
-    let valid_nodes = valid_nodes_itr.into_iter().collect::<BTreeSet<_>>();
-    let valid_voters = voters_itr
-        .into_iter()
-        .filter(|voter| valid_nodes.contains(voter))
-        .collect::<BTreeSet<_>>();
-
+pub fn quorum(voters: &BTreeSet<PublicInfo>, valid_nodes: &BTreeSet<PublicInfo>) -> bool {
     let valid_total_age = valid_nodes.iter().map(|node| usize::from(node.age())).sum();
     let mut running_total_age = 0;
-    for (count, voter) in valid_voters.iter().rev().enumerate() {
+
+    let valid_voters = voters.intersection(valid_nodes);
+    for (count, voter) in valid_voters.enumerate() {
```

Comment:
Yeah I know. Problem is that the `btree_set::Intersection` iterator is not a `DoubleEndedIterator` so it cannot be reversed. I would have to `collect` the iterator into something reversible and use that, and I'm not sure it would be worth it. Besides, `valid_voters` contains at most `GROUP_SIZE` elements, doesn't it? So the performance difference is going to be negligible I think.

Reply:
Ah - fair enough then!